### PR TITLE
Updates the directory sharing rfd to include a proposal for audit events

### PR DESCRIPTION
**Pull request by ibeckermayer**

### CODE REVIEW

**Review by zmb3 — COMMENTED** [review]
Seems reasonable to me. Only question is whether it is going to be spammy. For example, when copying a large file into or out of the shared directory, will we emit hundreds of read/write events? If so, we may want to consider debouncing, or otherwise detecting the last write and only emitting a single event.

**Comment by ibeckermayer** [comment]
> Seems reasonable to me. Only question is whether it is going to be spammy. For example, when copying a large file into or out of the shared directory, will we emit hundreds of read/write events?

How many read/write events are omitted is determined by the program on the Windows machine. In the case of copying files in/out of the shared directory that program is typically File Explorer, which empirically does large read/writes 1MB at a time (exactly 2^20 bytes at a time).

> If so, we may want to consider debouncing, or otherwise detecting the last write and only emitting a single event.

Debouncing is difficult to do because we don't have any reliable message coming through that tells us "this sequence of read/writes is done". RDP itself does have a "close file handle" mechanism that works approximately that way, but we handle it in the Rust code and don't translate it into a TDP message. (Now that I understand the system better I realize that it wouldn't be the worst idea to make TDP work more like RDP in this respect, but that's a much larger project).

A couple of options I can see that wouldn't require such a significant refactoring:

1. **Cache + new TDP message**: We cache every read and write near the event emitter, and add a new TDP message that's sent whenever we get an RDP "close file handle" mechanism. Upon receipt, we traverse through the cache and consolidate sequential read/writes for the given file into a single audit event.
2. **Cache + periodic job**: We cache every read and write near the event emitter, and add a function that runs on a timer which traverses the cache, consolidating sequential read/writes to the same file into a single audit event.

Between these two options, I prefer option 1. With option 2, its still possible for what from a human perspective looks like a single large read to get broken up into multiple audit events, if the timer happens to trigger at an inopportune time, whereas this won't happen with option 1 (assuming the Windows side program isn't doing anything weird). Option 1 is also a first step in the larger refactor I mention parenthetically above. The disadvantage of option 1 is that it's a bit more work to add a new TDP message.

It's also of note that by caching read/writes (both options), we open up a greater risk of them not making it into the audit log. IIRC there is some recording mode where the system freezes up if events are not being sent (although I can't seem to find it in the docs) -- such an airtight guarantee would be difficult if not impossible with an event caching system.

**Review by probakowski — APPROVED** [review]

**Comment by espadolini** [comment]
Do we really need exact offsets for all read and write ops? We could emit an event for "user has interacted with file" and then maybe emit one with some collected details, periodically/on close.

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 1]
```diff
@@ -594,3 +594,204 @@ The UX for this feature is discussed in [RFD 0058](https://github.com/gravitatio
...
+- A file was transferred from the local machine (from the shared directory) onto the remote Windows machine
+- etc
+
+However due to our limited visibility into the Windows side of the equation it is not possible for us to determine such events with precision. For example, if a user transfers a file
+from the Windows box into the shared directory, we see that as a `Shared Directory Create` sequence followed by a `Shared Directory Write` sequence. But then that same sequence could
+just have been a user creating a brand new file within the shared directory and then writing to it (which is not of particular note from a security perspective). Given this limitation,
+we will instead simply log all events that could conceivably indicate security-relevant information, and document the expected sequence of events for such common scenarios such as those
+described above for our users to reference. `Shared Directory Announce/Acknowledge` is also included in the audit event log to make the sequence of events more easily comprehensible.
+
+**TDP Shared Directory Messages Logged**
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:612]: Let's make this less about messages and more about events. For example, list the proposed event name and when it occurs.

- `windows.desktop.directory.share`: indicates that the user has shared a directory with the remote desktop
- `windows.desktop.directory.{read|write}`: etc...

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 2]
```diff
@@ -594,3 +594,204 @@ The UX for this feature is discussed in [RFD 0058](https://github.com/gravitatio
...
+The read/write events that are sent to us are fundamentally determined by the program on the Windows machine that's reading or writing to a shared file.
+In the case of copying files in/out of the shared directory, that program is typically File Explorer, which empirically does read/writes at a maximum of
+1MB (exactly 2^20 bytes) at a time. For files larger than 1MB, this manifests as several read/writes in a row showing up at time in such a case.
+
+In order to avoid the logs getting spammed with multiple messages corresponding to one "operation" (such as a multi MB read of a large file), we can create
+an algorithm which upon reciept of a read or write event, sets a short timer (say 1s), and if it receives another non-overlapping read/write for the same
+file before the timer runs out, stops the timer and amalgamates the events together, repeating until the timer runs out at which point a single event is
+written into the log (credit to @zmb3 for the algo design).
+
+This feature will be considered beyond the scope of the initial implementation, which will naively write all reads/writes as individual events.
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:797]: Fair. 👍

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 3]
```diff
@@ -594,3 +594,204 @@ The UX for this feature is discussed in [RFD 0058](https://github.com/gravitatio
...
+
+```proto
+// DesktopSharedDirectoryDelete is emitted when Teleport
+// attempts to delete a file or directory in a shared directory
+// at the behest of the remote desktop.
+message DesktopSharedDirectoryDelete {
+  // Metadata, UserMetadata, SessionMetadata, ConnectionMetadata ommitted
+
+  // Succeeded is true if the process was successful.
+  bool Succeeded = 5 [(gogoproto.jsontag) = "succeeded"];
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:752]: Can you leverage `message Status` here (and below)?

**Comment by ibeckermayer** [comment]
@espadolini 

> Do we really need exact offsets for all read and write ops? We could emit an event for "user has interacted with file" and then maybe emit one with some collected details, periodically/on close.

The idea with including offsets is to provide an auditor with as much information as possible without actually logging file data. Your proposal would work too, but requires a more complex implementation for not much of a UX or other benefit as far as I'm able to discern.

File: rfd/0067-desktop-access-file-system-sharing.md (hunk 1 above)
Comment (ibeckermayer) [reply rfd/0067-desktop-access-file-system-sharing.md]: I went with `"desktop.directoy.*"` based on other events here
https://github.com/gravitational/teleport/blob/f87519333620dd5105b376ff2edfc8b71279080d/lib/events/api.go#L545-L552

Though I noticed that our start and end events are of the `"windows.desktop.*"` form. Perhaps we should modify one or the other to stick to a single pattern?

https://github.com/gravitational/teleport/blob/f87519333620dd5105b376ff2edfc8b71279080d/lib/events/api.go#L528-L533

**Comment by espadolini** [comment]
The benefit would be to reduce audit log spam while still guaranteeing that at least some info gets in even if Teleport stops abruptly.

**Comment by ibeckermayer** [comment]
> The benefit would be to reduce audit log spam while still guaranteeing that at least some info gets in even if Teleport stops abruptly.

Maybe I misunderstood. I thought you were saying to just emit a generic "interacted" whenever we see any of these various actions (read/write/move/delete/etc), and then later emit a summary, but that would be the same amount of events. But I suppose instead you're suggesting that we just add a single event for interaction with any particular file, and then emit a summary.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**